### PR TITLE
prevent `allocMappedBufIfSupported` from accepting arbitrary Types as `TPlatform`

### LINKED ISSUE #2065: `allocMappedBufIfSupported` silently accepts non-platform types as TPlatform
[`alpaka::allocMappedBufIfSupported`](https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/mem/buf/Traits.hpp#L159-L188) currently does not directly check if TPlatform is an actual platform type (`alpaka::isPlatform<TPlatform>`).

This can lead to cases where an inexperienced user can pass the device type instead of the platform type as TPlatform to `alpaka::allocMappedBufIfSupported`. Since `alpaka::hasMappedBufSupport` is [false by default](https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/mem/buf/Traits.hpp#L43) for every type, the [second if-branch within allocMappedBufIfSupported](https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/mem/buf/Traits.hpp#L183-L185) is chosen, and non-Managed memory is returned even when managed memory would be available. All this happens silently, without giving the user a warning that TPlatform is not a Platform, which can be very confusing.

To prevent this, [this implementation](https://github.com/chaever/alpaka/blob/check_TPltf_when_allocating_device_memory/include/alpaka/mem/buf/Traits.hpp#L178) includes a `static_assert` to notify the user in this case.

### PR DESCRIPTION
**Pull request by chaever**
This PR aims to resolve #2065.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
@chaever Thanks for your first PR. We will have a look into it!

File: include/alpaka/mem/buf/Traits.hpp [hunk 1]
```diff
@@ -175,9 +176,10 @@ namespace alpaka
         TPlatform const& platform,
         TExtent const& extent = TExtent())
     {
-        if constexpr(hasMappedBufSupport<TPlatform>)
+        using Platform = Platform<TPlatform>;
```
Comment (psychocoderHPC) [inline include/alpaka/mem/buf/Traits.hpp]: CI error: 

```
/home/runner/work/alpaka/alpaka/include/alpaka/mem/buf/Traits.hpp:179:26: error: 'Platform' does not name a type; did you mean 'TPlatform'?
  179 |         using Platform = Platform<TPlatform>;
      |                          ^~~~~~~~
      |                          TPlatform
```

The include to `alpaka::platform/Traits.hpp` is missing.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/mem/buf/Traits.hpp [hunk 2]
```diff
@@ -1,4 +1,5 @@
-/* Copyright 2023 Alexander Matthes, Benjamin Worpitz, Andrea Bocci, Bernhard Manfred Gruber, Jan Stephan
+/* Copyright 2023 Alexander Matthes, Benjamin Worpitz, Andrea Bocci, Bernhard Manfred Gruber, Jan Stephan,
+ * Christian Kaever
```
Comment (j-stephan) [inline include/alpaka/mem/buf/Traits.hpp]:
```suggestion
/* Copyright 2023 Alexander Matthes, Benjamin Worpitz, Andrea Bocci, Bernhard Manfred Gruber, Jan Stephan,
 *                Christian Kaever
```

File: include/alpaka/mem/buf/Traits.hpp (hunk 1 above)
Comment (j-stephan) [reply include/alpaka/mem/buf/Traits.hpp]: Won't `Platform = Platform<TPlatform>` also trigger some kind of shadowing warning?

**Review by j-stephan — CHANGES_REQUESTED** [review]

**Review by j-stephan — DISMISSED** [review]
Thank you!

File: include/alpaka/mem/buf/Traits.hpp (hunk 1 above)
Comment (chaever) [reply include/alpaka/mem/buf/Traits.hpp]: I've changes the name to `PlatformForMappedAlloc` (which is used nowhere else currently).
Is this okay?

Comment (j-stephan) [reply include/alpaka/mem/buf/Traits.hpp]: Yes, that's fine.

File: include/alpaka/mem/buf/Traits.hpp [hunk 3]
```diff
@@ -175,9 +177,10 @@ namespace alpaka
         TPlatform const& platform,
         TExtent const& extent = TExtent())
     {
-        if constexpr(hasMappedBufSupport<TPlatform>)
+        using PlatformForMappedAlloc = Platform<TPlatform>;
```
Comment (fwyzard) [inline include/alpaka/mem/buf/Traits.hpp]: I would suggest
```suggestion
        using Platform = alpaka::Platform<TPlatform>;
```
if it works?

Comment (fwyzard) [reply include/alpaka/mem/buf/Traits.hpp]: See for example https://github.com/alpaka-group/alpaka/blob/ecf917fa2a7d4cfa6e15e827688193c5e77302d5/include/alpaka/test/queue/QueueTestFixture.hpp#L17

Comment (chaever) [reply include/alpaka/mem/buf/Traits.hpp]: This should now be consistent with the other usages of the `Platform`-Trait, which all have the `alpaka` namespace prefixed.

File: include/alpaka/mem/buf/Traits.hpp [hunk 4]
```diff
@@ -175,9 +177,10 @@ namespace alpaka
         TPlatform const& platform,
         TExtent const& extent = TExtent())
     {
-        if constexpr(hasMappedBufSupport<TPlatform>)
+        using PlatformForMappedAlloc = alpaka::Platform<TPlatform>;
```
Comment (fwyzard) [inline include/alpaka/mem/buf/Traits.hpp]: can you use `using Platform` like everywhere else?

Comment (chaever) [reply include/alpaka/mem/buf/Traits.hpp]: That was the initial version, until [this](https://github.com/alpaka-group/alpaka/pull/2120#discussion_r1312843301) was pointed out.

Comment (j-stephan) [reply include/alpaka/mem/buf/Traits.hpp]: My assumption was wrong. `Platform = Platform<T>` doesn't trigger `-Wshadow`. So feel free to follow @fwyzard's suggestion.

Comment (chaever) [reply include/alpaka/mem/buf/Traits.hpp]: `PlatformForMappedAlloc` should now be replaced by `Platform`.

**Review by psychocoderHPC — APPROVED** [review]

**Review by j-stephan — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
@chaever  Could you please call `git commit --amend   -a` on you terminal and force push again. We had last week CI issues and must retrigger the CI.

**Comment by chaever** [comment]
I think the same [job](https://gitlab.com/hzdr/crp/alpaka/-/jobs/5021314057) failed again.
It looks like that has something to do with a failure to transfer the code for test compilation, see [here](https://gitlab.com/gitlab-org/gitlab/-/issues/360559).

**Comment by j-stephan** [comment]
Can you do a `git rebase develop` and then force-push again?

**Comment by chaever** [comment]
This branch should now be branching off from develop at 6b43b8d (most recent develop-commit at time of writing).

**Pull request merged**